### PR TITLE
Support `createdBy=me` for all endpoints that use the `createdBy` filter

### LINKED ISSUE #960: Create a GET `/bulkUploads?mine` parameter
We need to be able to load the bulk uploads for the current user; this is technically possible to do from a client, but it involves looking up the current user and passing that user's ID into the `createdBy` parameter.

We'd like to create a new `mine` (name tbd) boolean, which will only return data owned by the current user.  This would override any `createdBy` filter.

Given the pressure from the demo timeline this seems like a reasonable API feature.  We may want to reflect on it later to make sure it's appropriate to keep in the long term.

### PR DESCRIPTION
**Pull request by slifty**
This PR lets you say `createdBy=me` to retrieve your own data in any instance where a `createdBy` parameter is accepted.

Resolves #960

### CODE REVIEW

File: src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+	});
+
+	it('should pass along the numeric value when one is provided', () => {
+		const paginationParameters = extractCreatedByParameters({
+			query: {
+				createdBy: '42',
+			},
+		});
+		expect(paginationParameters).toEqual({
+			createdBy: 42,
```
Comment (reefdog) [inline src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts:22]: wow this is elegant

File: src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		expect(createdByParameters).toEqual({
+			createdBy: testUser.id,
+		});
+	});
+
+	it('should throw an error when strings that parse to NaN are provided', () => {
+		expect(() =>
+			extractCreatedByParameters({
+				query: {
+					createdBy: 'forty two',
```
Comment (reefdog) [inline src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts:54]: have i told you lately that you're elegant

File: src/queryParameters/extractCreatedByParameters.ts [hunk 3]
```diff
@@ -2,35 +2,57 @@ import { ajv } from '../ajv';
...
+	createdBy: number | 'me' | undefined;
+}
 
 interface CreatedByParameters {
 	createdBy: number | undefined;
 }
 
-const createdByParametersSchema: JSONSchemaType<CreatedByParameters> = {
-	type: 'object',
-	properties: {
```
Comment (reefdog) [inline src/queryParameters/extractCreatedByParameters.ts:12]: removing this line is very elegant

File: src/queryParameters/extractCreatedByParameters.ts [hunk 4]
```diff
@@ -2,35 +2,57 @@ import { ajv } from '../ajv';
...
+			isCreatedByQueryParameters.errors ?? [],
 		);
 	}
+
+	const createdBy =
+		query.createdBy === 'me' ? partialAuthContext?.user.id : query.createdBy;
+
 	return {
-		createdBy: query.createdBy,
+		createdBy,
```
Comment (reefdog) [inline src/queryParameters/extractCreatedByParameters.ts:55]: this is legitimately a more elegant syntax

File: src/queryParameters/extractCreatedByParameters.ts [hunk 5]
```diff
@@ -2,35 +2,57 @@ import { ajv } from '../ajv';
...
+						enum: ['me'],
+					},
+				],
+				nullable: true,
+			},
 		},
-	},
-	required: [],
-};
+		required: [],
```
Comment (reefdog) [inline src/queryParameters/extractCreatedByParameters.ts:34]: is there any way to make this more elegant?

File: src/queryParameters/extractCreatedByParameters.ts [hunk 6]
```diff
@@ -2,35 +2,57 @@ import { ajv } from '../ajv';
 import { InputValidationError } from '../errors';
 import type { JSONSchemaType } from 'ajv';
 import type { Request } from 'express';
+import type { AuthContext } from '../types';
```
Comment (reefdog) [inline src/queryParameters/extractCreatedByParameters.ts:5]: this is very elephant

**Comment by reefdog** [comment]
Okay genuinely this doesn't _seem_ to have an effect. At least, when I call `/bulkUploads?createdBy=me` (via front end, Swagger, cURL, whatever), it's the same as calling `/bulkUploads`. I get the full set either way. (And yes, the full set has uploads from multiple users. `/bulkUploads?createdBy=1` returns just 10, as expected.)

**Comment by slifty** [comment]
Justin I didn't ask for your review yet

**Comment by slifty** [comment]
ok @reefdog could you please review.

File: src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts (hunk 1 above)
Comment (slifty) [reply src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts:22]: UNSUBSCRIBE.

File: src/queryParameters/extractCreatedByParameters.ts (hunk 4 above)
Comment (slifty) [reply src/queryParameters/extractCreatedByParameters.ts:54]: OK Justin. But listen. Isn't it kind of amazing that we expanded the functionality of TWO endpoints without even editing EITHER of them!?

File: src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	it('should throw an error when strings that parse to numbers less than 1 are provided', () => {
+		expect(() =>
+			extractCreatedByParameters({
+				query: {
+					createdBy: '0',
+				},
+			}),
+		).toThrow(InputValidationError);
+	});
+});
```
Comment (reefdog) [inline src/queryParameters/__tests__/extractCreatedByParameters.int.test.ts:75]: This is a really good set of tests.

File: src/queryParameters/extractCreatedByParameters.ts (hunk 4 above)
Comment (reefdog) [reply src/queryParameters/extractCreatedByParameters.ts:54]: Yes!

File: src/__tests__/bulkUploads.int.test.ts [hunk 8]
```diff
@@ -184,6 +184,46 @@ describe('/bulkUploads', () => {
 				);
 		});
 
+		it('returns uploads for the admin user when createdBy is set to me as an admin', async () => {
```
Comment (reefdog) [inline src/__tests__/bulkUploads.int.test.ts:187]: Out of curiosity, why are these tests framed around admins? It implies to me this functionality is meant to only work for admins, when that's not the case. And in fact I think it'd be good to have tests that ensure this *isn't* somehow an admin-only feature. Right?

I'm not insistent on this point, and maybe there's a good reason!

**Review by reefdog — APPROVED** [review]
Huzzah! This is very elegant. A few questions and observations sprinkled about, but I'm happy with this and grateful you knocked it out.

File: src/__tests__/bulkUploads.int.test.ts (hunk 8 above)
Comment (slifty) [reply src/__tests__/bulkUploads.int.test.ts:187]: Main reason right now is that admin is the only role where the filter matters.  If you aren't an admin you only get your own things anyway.

"admin" in the test name is just placeholder for "user where more than just their stuff would have been returned in the first place"

**Pull request merged**